Script Cleanup

Foggalong opened the pull request:
This PR fixes a couple of bugs with the script part of Core, increases efficiency, and generally cleans up the codebase. Lots of the code hasn't been touched since it was first written because it Just Works™, but for that reason some issues have gone unchecked.

I'll comment on the commits themselves with specific improvement details, but in short this is what this PR hopes to achieve:
* [x] Complete review of the test scripts
* [x] Complete review of the main generator code
* [x] Fixes #3453
* [x] Fixes #4047
* [x] Fixes #4779 
* [x] Fixes #4786
* [x] Add action for checking `.editorconfig` followed

Foggalong committed:
Clean up test scripts

Foggalong commented on tests/db_entry.py:
```diff
@@ -1,34 +1,34 @@
-# published by the Free Software Foundation. You should have received
-# a copy of the GNU General Public License along with this program.
-# If not, see <http://www.gnu.org/licenses/>.
+Copyright (C) 2019 Numix Project
```
Updated all copyright headers, removed unneeded `#` characters

Foggalong commented on tests/db_entry.py:
```diff
@@ -1,34 +1,34 @@
 
+from json import load
 from os import path
-import json
```
Alphabetised all (global) import lists

Foggalong commented on tests/db_entry.py:
```diff
@@ -1,34 +1,34 @@
-import json
 
-from utils import error
+from utils import error, THEMES
```
The list of themes was used in multiple tests so made it part of `utils.py`

Foggalong commented on tests/db_entry.py:
```diff
@@ -1,34 +1,34 @@
 ICONS_DIR = path.join(ABS_PATH, "../icons/")
 
+with open(DB_FILE, 'r') as db_obj:
+    data = load(db_obj)
```
The actual test is no longer carried out inside the file-read `with` block.

Foggalong commented on tests/db_entry.py:
```diff
@@ -1,34 +1,34 @@
+    for theme in THEMES:
+        icon = path.join(ICONS_DIR, theme, "48/{}.svg".format(entry))
+        if path.exists(icon):
+            continue
```
Using a `continue` rather than an extra level of indentation after a lone-`if` makes the code cleaner, especially later one when it's reaching 6-7 levels.

Foggalong commented on tests/db_entry.py:
```diff
@@ -1,34 +1,34 @@
+        has_errors = True
+        error("'{}' doesn't have an icon in the {} theme".format(entry, theme))
+
+exit(has_errors)
```
`has_errors` is a bool value so it doesn't need converting to an integer before being passed to `exit()`

Foggalong commented on tests/icon_cache.py:
```diff
@@ -1,41 +1,39 @@
 DB_FILE = path.join(ABS_PATH, "../data.json")
 
 with open(DB_FILE, 'r') as db_obj:
-    data = json.load(db_obj, object_pairs_hook=OrderedDict)
```
This test loaded the JSON file as an ordered dictionary which isn't needed here.

bilelmoussaoui replied:
Are you sure about this? if you load the json file here, it will result in an unordered dict.

Foggalong replied:
The point was more that the `icon_cache` test doesn't need to load the JSON in a special way, the standard load used elsewhere is fine. I didn't test if `object_pairs_hook=OrderedDict` ordered the dictionary, I just assumed it did based on the name.

bilelmoussaoui replied:
sorry I misunderstood what you said then :p it's perfect! 
the object_pairs_hook uses a new object for each child on your json object, as dict don't preserve the order, using OrderDict makes it use the same order as the one on the file

Foggalong replied:
Ahhhh, I get you. That actually makes a lot of sense, thanks! Also if anything else about this pull catches your eye as iffy let me know, I realise it's changing a lot of code you originally wrote

bilelmoussaoui replied:
You can ping me here whenever you need that. Will take a closer look once you think it should be merged ^^

Foggalong commented on tests/icon_cache.py:
```diff
@@ -1,41 +1,39 @@
-from collections import OrderedDict
 
-from utils import error
+from utils import error, space_check
```
`space_check` is now a `utils.py` function replacing `test_empty_space`

Foggalong commented on tests/db_entry.py:
```diff
@@ -1,34 +1,34 @@
+            continue
+
+        has_errors = True
+        error("'{}' doesn't have an icon in the {} theme".format(entry, theme))
```
Added quotation marks `'...'` around DB entries in all cases, rather than speech marks `"..."` around only some cases.

Foggalong commented on tests/icons_entry.py:
```diff
@@ -1,38 +1,38 @@
-                      icon_name
-                  ))
-exit(int(has_errors))
+        if icon_name in entries or icon_name in reported:
```
As well as the previous comment about `continue` being cleaner than a whole new `if` level, this has an efficiency improvement. For those who know logic, you'll recognise that:

```
(¬A) ʌ (¬B) === ¬(A v B)
```

Since the continue handles the `NOT` on the right hand side, we're actually checking whether just `(A v B)`. Since `A` and `B` are just inclusion events (`x in X`) they're much quicker than evaluating the previous exclusion events (`x not in X`). This is further improved because Python halts logical evaluations as soon as it knows the statement as a whole is true, i.e. here if `icon_name in entries == True` it doesn't bother evaluating `icon_name in reported` and continues to the `continue`.

Foggalong commented on tests/ordered_db.py:
```diff
@@ -1,57 +1,40 @@
+from os import path
 
-from utils import sort, error
+from utils import error, sort, sort_check
```
The code for checking whether a list was sorted was included three times separately in this script rather than being turned into a function. It's now done as part of `utils.py`.

Foggalong commented on tests/ordered_db.py:
```diff
@@ -1,57 +1,40 @@
+has_errors = sort_check(list(data.keys()), "Database entry")
 
 for key, value in data.items():
+    if value.get("android"):
```
Moved Android to top to match JSON.

Foggalong commented on tests/ordered_db.py:
```diff
@@ -1,57 +1,40 @@
+            continue
+        has_errors = sort_check(symlinks, "Linux symlink", key)
+
+    if value.get("osx"):
```
Added OSX check.

Foggalong commented on tests/schema.json:
```diff
@@ -4,6 +4,14 @@
+                    },
+                    "minItems": 1,
+                    "uniqueItems": true
+                },
```
Moved Android to top to match `data.json`

Foggalong commented on tests/schema.json:
```diff
@@ -26,7 +34,7 @@
                     ]
                 },
-                "android": {
+                "osx": {
```
Added OSX

Foggalong commented on tests/schema.py:
```diff
@@ -1,23 +1,25 @@
-DB_FILE = path.join(path.dirname(path.abspath(__file__)), "../data.json")
-SCHEMA_FILE = path.join(path.dirname(path.abspath(__file__)), 'schema.json')
+
+ABS_PATH = path.dirname(path.abspath(__file__))
```
Used twice so made a variable

Foggalong commented on tests/utils.py:
```diff
@@ -23,3 +21,43 @@ def error(msg):
+        error("'{}' contains an empty space".format(string))
+        return True
+    else:
+        return False
```
Returns a bool rather than making use of global variables as previous.

Foggalong committed:
Standalone sort() not needed

Foggalong commented on tests/utils.py:
```diff
@@ -1,18 +1,16 @@
-def sort(icons_list):
-    """Sort list case insensitive."""
-    return sorted(icons_list,
-                  key=lambda icon_name: icon_name.lower())
```
Deleted `sort()` since was only used in `sort_check()`

Foggalong commented on tests/utils.py:
```diff
@@ -23,3 +21,38 @@ def error(msg):
+    optional string 'root' as a root value for use with sublists.
+    """
+    has_errors = False
+    reference = sorted(list, key=lambda item: item.lower())
```
The previous `sort` function absorbed into here.

Foggalong commented on tests/utils.py:
```diff
@@ -23,3 +21,38 @@ def error(msg):
+        else:
+            error_msg = "{} '{}' of '{}' not correctly ordred"
+
+        error(error_msg.format(stype, list[i], root))
```
Extraneous `format` arguments are simply ignores so using one `error` for both messages fine.

Foggalong commented:
I wouldn't normally thoroughly review my own code changes like this but I didn't actually write any of these tests, @bilelmoussaoui did. With that in mind I thought it only fair I explain what's being changed.

Foggalong committed:
Generation script clean-up

Foggalong commented on gen.py:
```diff
@@ -18,33 +18,35 @@
 try:
     from gi import require_version
-    require_version('Rsvg', '2.0')
+    require_version("Rsvg", "2.0")
```
Switched `'` for `"` for consistency throughout. The only exceptions are strings for `app_filer` because that uses `"` characters already as part of its content.

Foggalong commented on gen.py:
```diff
@@ -18,33 +18,35 @@
-    else:
+    # Assigns true if Inkscape found, false otherwise
+    use_inkscape = not call(["which", "inkscape"], stdout=PIPE, stderr=PIPE)
+    if not use_inkscape:
```
Rather than using distinct `flag` and `use` variables, can just use one with logical `not` converters.

Foggalong commented on gen.py:
```diff
@@ -18,33 +18,35 @@
-parser = ArgumentParser(prog="Numix-core")
-
-# Importing JSON
+# Loading the JSON data file
```
Load JSON before starting work with the parser

Foggalong commented on gen.py:
```diff
@@ -55,15 +57,17 @@
-    exit("Please use --theme argument with "
-         "one of the following: {}".format(", ".join(themes)))
+    msg = "Please use --theme argument with one of the following: "
+    exit(msg + ", ".join(themes))
```
There were quite a few cases where `format` wasn't needed; standard string concatenation sufficed so uses that.

Foggalong commented on gen.py:
```diff
@@ -97,93 +101,119 @@ def convert_svg2png(infile, outfile, w, h):
-    sizes = listdir("icons/{0}".format(theme))
-except FileNotFoundError:
-    exit("The theme {0} does not exists. Please reclone"
-         "the repository and try again.".format(theme))
```
Because of the prior `try` block in line 42-43 this `try block` (previously lines 107-112) can never actually be trigger. That is, except if the theme directory is deleted between the previous check and this one. Covering that gets us into race conditions though which is a path I'd sooner not go down.

Foggalong commented on gen.py:
```diff
@@ -97,93 +101,119 @@ def convert_svg2png(infile, outfile, w, h):
-            app_filter_content += '\t<item component="ComponentInfo{' + \
-                component_info + '}" drawable="' + drawable_name + '"/>\n'
+            output = "{0}/{1}.png".format(theme_dir, drawable_name)
+            convert_svg2png(source, output, 192, 192)
```
This was actually a bug! Previously, the icon was added to the app_filter file even if it didn't have an icon created for it.

Foggalong commented on gen.py:
```diff
@@ -97,93 +101,119 @@ def convert_svg2png(infile, outfile, w, h):
+
+            # Add icons app_filter data (afd) to the app_filter file
+            afd = '\t<item component="ComponentInfo{{{0}}}" drawable="{1}"/>\n'
+            app_filter_content += afd.format(component_info, drawable_name)
```
Cleaner representation of what this line actually looks like than splitting it as previous.

Foggalong commented on gen.py:
```diff
@@ -97,93 +101,119 @@ def convert_svg2png(infile, outfile, w, h):
+
+    # OSX generation depends on libicns function png2icns
+    no_icns = call(["which", "png2icns"], stdout=PIPE, stderr=PIPE)
+    if no_icns:
```
Renamed variable, pretty sure this was just a copy-paste error.

Foggalong commented on gen.py:
```diff
@@ -97,93 +101,119 @@ def convert_svg2png(infile, outfile, w, h):
+            # Icon output locations
+            svg_icon = "{0}/vectors/{1}.svg".format(osx_dir, output_icon)
+            png_icon = "{0}/pngs/{1}.png".format(osx_dir, output_icon)
+            icn_icon = "{0}/icns/{1}.icn".format(osx_dir, output_icon)
```
Not gonna lie, I renamed these variables *entirely* so that the final `call` didn't have to be split over two lines.

Foggalong commented:
Something tweaked in the interim has fixed #4779 - my guess is that it was just a case of a character not being being correctly escaped but is as part of some formatting changes.

Foggalong committed:
Made PEP8 job, not a build step

Will run in parallel so should be slightly quicker, removes need for bash file

Foggalong committed:
Trialling a less agressive install strategy

Foggalong committed:
Added action for markdown lint

Foggalong committed:
Added .editorconfig action

Foggalong committed:
Fixed existing linter errors

Foggalong committed:
Exclude license from MD lint

Foggalong committed:
Removed JSON rule, added SVG rule

Foggalong committed:
Added duplicate icon name test
Fixes #4786

Foggalong committed:
Remove duplicates from data.json

Foggalong committed:
Refine `utils.py` usage
Define path variables in utils, move single use functions out of utils, and add docstrings to all the tests

Foggalong committed:
Adds icon tags - fixes #3453

Adds icon tags to group related icons and a script for extracting them to a separate directory. Can be used either for related icons (e.g. ms-office,
icons using the same design element (e.g. world-baseplate), etc.

Foggalong committed:
Fixed tags.py being picked up as a theme

Foggalong committed:
Switch from pycodestyle to flake8

Foggalong commented:
It's just suddenly dawned on me that most the dependencies that have been added in `test.yml` and `test-requirements.txt` aren't actually test dependencies, they're _build_ dependencies. We should be able to run the tests just fine without installing anything extra except `flake8` and `jsonschema`.

If we later add a GitHub action to automatically build and push to `numix-icon-theme-*` then those will need them other dependencies, but that will also have to be a separate workflow since we wouldn't want it to trigger `on: pull_request`

Foggalong committed:
Removed unnecessary dependencies
